python/cuml/cuml/internals/interop.py

```diff
@@ -0,0 +1,311 @@
+        else:
+            return gpu_func(self, *args, **kwargs)
+
+    return inner
+
+
+def to_gpu(x, order="K"):
+    """Coerce `x` to the equivalent GPU type."""
+    # Ideally we'd get rid of this helper eventually, using it for now to reduce typing
+    from cuml.internals.input_utils import input_to_cuml_array
+
+    return input_to_cuml_array(
+        x, order=order, convert_to_mem_type=MemoryType.device
+    )[0]
+
+
+class UnsupportedOnGPU(ValueError):
+    """An exception raised when a conversion of a CPU to a GPU estimator isn't supported"""
+
+
+class UnsupportedOnCPU(ValueError):
+    """An exception raised when a conversion of a GPU to a CPU estimator isn't supported"""
+
+
+class classproperty:
+    """Like ``property``, but on a class rather than an instance."""
+
+    def __init__(self, fget):
+        self.fget = fget
+
+    def __get__(self, obj, owner):
+        return self.fget(owner)
+
+
+class InteropMixin:
+    """A mixin for enabling conversion of a cuml estimator to/from its
+    CPU-based counterpart.
+
+    Subclasses should define:
+
+    - `_cpu_class_path`
+    - `_params_from_cpu`
+    - `_params_to_cpu`
+    - `_attrs_from_cpu`
+    - `_attrs_to_cpu`
+
+    In return they get ``as_sklearn``/``from_sklearn`` methods, and can also
+    be used as the basis for a proxy estimator in ``cuml.accel``.
+    """
+
+    # The import path to use to import the CPU model class
+    _cpu_class_path: str
+
+    @classproperty
+    def _cpu_class(cls):
+        """The CPU class that corresponds to this GPU model"""
+        module, _, name = cls._cpu_class_path.rpartition(".")
+        return getattr(import_module(module), name)
+
+    @classmethod
+    def _params_from_cpu(cls, model) -> dict[str, Any]:
+        """Get parameters to use to instantiate a GPU model from a CPU model.
+
+        Parameters
+        ----------
+        model
+            The CPU model, of the same type as ``self._cpu_class``.
+
+        Returns
+        -------
+        dict
+            A mapping of keyword arguments that may be used to instantiate ``cls``
+            to create an equivalent GPU model.
+
+        Raises
+        ------
+        UnsupportedOnGPU
+            If one or more hyperparameters are unsupported by the GPU model.
+        """
+        raise NotImplementedError
+
+    def _params_to_cpu(self) -> dict[str, Any]:
+        """Get parameters to use to instantiate a CPU model from a GPU model.
+
+        Returns
+        -------
+        dict
+            A mapping of keyword arguments that may be used to instantiate
+            ``self._cpu_class`` to create an equivalent CPU model.
+
+        Raises
+        ------
+        UnsupportedOnCPU
+            If one or more hyperparameters are unsupported by the CPU model.
+        """
+        raise NotImplementedError
+
+    def _attrs_from_cpu(self, model) -> dict[str, Any]:
+        """Get attributes to set on ``self`` from a fit CPU model.
```

Comment:
Should this method return all attributes that it wants to have set on the GPU model or just "fitted attributes" (those ending in a `_`)?

Leaving the door open for this to return "any and all" attributes, maybe even ones that don't exist on the CPU model at all but need to be set on the GPU model could be interesting. I don't have a use-case in mind, but maybe the fact that `model.solver=="lbfgs"` means that you want to return `{"_solver": cuml.solvers.Foo()}` from this method.

Reply:
Ideally we're doing "fitted attributes" which match close to 1:1 with the state sklearn stores (and anything private is handled via properties etc...). That said, really the `_params_*` methods are for parameters only and `_attrs_*` is for anything else necessary. Hopefully as we move things out of `__init__` and consolidate state on estimators the `_attrs_*` methods become simple and mostly moving fitted attributes.